v4l2/subdev.py

```diff
@@ -271,3 +274,53 @@ def set_control(self, ctrl_id: int, ctrl_val: int):
         v4l2_ctrl.value = ctrl_val
 
         fcntl.ioctl(self.fd, v4l2.uapi.VIDIOC_S_CTRL, v4l2_ctrl, False)
+
+    def enum_ext_ctrls(self):
+        ctrl = v4l2.uapi.v4l2_queryctrl()
+        ctrl.id = v4l2.uapi.V4L2_CTRL_FLAG_NEXT_CTRL
+        ext_ctrls = {}
+        while True:
+            try:
+                fcntl.ioctl(self.fd, v4l2.uapi.VIDIOC_QUERYCTRL, ctrl)
+            except OSError:
+                return ext_ctrls
+
+            if ctrl.type != v4l2.uapi.V4L2_CTRL_TYPE_CTRL_CLASS:
+                ext_ctrls[ctrl.id] = copy(ctrl)
+
+            ctrl.id |= v4l2.uapi.V4L2_CTRL_FLAG_NEXT_CTRL
+
+    def get_ext_ctrl(self, ctrl_id: int):
+        ctrl = v4l2.uapi.v4l2_ext_control()
+        ctrl.id = ctrl_id
+        ext_ctrl = v4l2.uapi.v4l2_ext_controls()
+        ext_ctrl.which = v4l2.uapi.V4L2_CTRL_WHICH_CUR_VAL
+        ext_ctrl.count = 1
+        ext_ctrl.controls = ctypes.pointer(ctrl)
+        fcntl.ioctl(self.fd, v4l2.uapi.VIDIOC_G_EXT_CTRLS, ext_ctrl)
+        return (
+            ctrl.value64
+            if (self.controls[ctrl_id].type == v4l2.uapi.V4L2_CTRL_TYPE_INTEGER64)
+            else ctrl.value
+        )
+
+    def set_ext_ctrl(self, ctrl_id, ctrl_val):
+        # TODO: handle situation when control is dynamically grabbed
+        if self.controls[ctrl_id].flags & v4l2.uapi.V4L2_CTRL_FLAG_READ_ONLY:
+            print(f"Control id [{ctrl_id}] isn't writable")
+            return
+        if self.controls[ctrl_id].flags & v4l2.uapi.V4L2_CTRL_FLAG_GRABBED:
+            print(f'Control id [{ctrl_id}] temporarily grabbed')
+            return
+
+        ctrl = v4l2.uapi.v4l2_ext_control()
+        ctrl.id = ctrl_id
+        if self.controls[ctrl_id].type == v4l2.uapi.V4L2_CTRL_TYPE_INTEGER64:
+            ctrl.value64 = ctrl_val
+        else:
+            ctrl.value = ctrl_val
+        ext_ctrl = v4l2.uapi.v4l2_ext_controls()
+        ext_ctrl.which = v4l2.uapi.V4L2_CTRL_WHICH_CUR_VAL
+        ext_ctrl.count = 1
+        ext_ctrl.controls = ctypes.pointer(ctrl)
+        fcntl.ioctl(self.fd, v4l2.uapi.VIDIOC_S_EXT_CTRLS, ext_ctrl)
```

Comment:
It would help me to understand what you need if you would describe in the commit desc what you use this for (similarly for exporting CaptureStreamer). Looks like you just use it to set/get a single 32/64 bit value?

I'm not super familiar with v4l2 controls as I haven't used them too much, but I think I'd really like to be able to expose a unified API for the users. No "ext" vs "non-ext" stuff, but just a way to get and set controls, regardless if the driver supports ext or not.

So, we already have the "naive" set_control(), I'd like to see it extended so it just works, instead of adding set_ext_ctrl(), requiring the caller to know which one to use.

Reply:
Well, As explained above, I have a WIP 'mediapipe' branch (https://github.com/sylpe/pyv4l2/tree/mediapipe). 
On top of this I have some pytests which allows me to tests camera driver : mostly iterating over all the formats (width, heigth, pixelformat), and play with controls.

Please find below an example of how I use it : 

```
@pytest.mark.parametrize('duration', [2, 10], ids=duration_name)
@pytest.mark.parametrize('tp', [VD56G3_TP_HGREY, VD56G3_TP_VGREY], ids=tp_name)
def test_stream_tp(media_conf, duration, tp, request):
    """
    Test that the sensor stream correctly with Grey gradient Test Patterns.

    Ensure the output framerate is correct.
    Ensure the obtained image fit with format width and height.
    Ensure the orientation of the test pattern comply with the one configured.
    Save last frame as an image (for human inspection).
    """
    # Init : reset frame holder
    frames_holder.clear()
    # Init : setup media pipe
    media_pipe = MediaPipe(media_conf)
    media_pipe.setup_links()
    media_pipe.configure_subdevs()
    # Init : reset/set camera subdevice controls
    camera_subdev = media_pipe.subdevices[media_conf.get_src_entity().name]
    reset_camera_subdev_ctrls(camera_subdev)
    camera_subdev.set_ext_ctrl(v4l2.uapi.V4L2_CID_TEST_PATTERN, tp)
    # Init : gather test values
    (width, height, _) = get_subdev_source_fmt(media_conf)
    expected_fps = get_theorical_fps(camera_subdev)

    # Stream for 'duration' secs
    media_pipe.start_stream(on_frame_cb=collect_frame)
    sleep(duration)
    media_pipe.stop_stream()

    # Normalize collected frame in np arrays
    normalize_frames_holder()

    # Analyze output : framerate
    assert duration * expected_fps - 2 <= len(frames_holder) <= duration * expected_fps + 2
    # Analyze output : image format (width, height)
    assert frames_holder[0][2] == width
    assert frames_holder[0][3] == height
    # Analyze output : check images orientation (for first and last frame captured)
    orientation, magnitude = avr_gradient_direction(frames_holder[0][0])
    assert orientation <= 0.01 if tp == VD56G3_TP_HGREY else 89.99 < orientation < 90.01
    orientation, magnitude = avr_gradient_direction(frames_holder[-1][0])
    assert orientation <= 0.01 if tp == VD56G3_TP_HGREY else 89.99 < orientation < 90.01
    # Save last image present in the frame_holder
    test_name = request.node.name[5:]
    save_to_file(test_name, -1)
```

This could help understand better the context.

I would have like to use the existing set_control(), but for some of the controls (especially integer64) and flags, It was possible to go through the "legacy" control API.

I'm totally aligned with you, better to improve the existing API. Depending of your thoughts I could proposed a better implementation fitting with pyv4l2 philosophy.